config/http_config.go

```diff
@@ -713,26 +786,80 @@ func updateRootCA(cfg *tls.Config, b []byte) bool {
 	return true
 }
 
+// CertGetter defines how to access certificates.
+type CertGetter interface {
+	// GetCert returns the corresponding certificate, whether or not
+	// it's been updated, or an error if it's not possible to
+	// retrieve the certificate.
+	GetCert() (cert []byte, updated bool, err error)
+}
+
+// newCertStore creates a CertGetter that provides access to the
+// certificate stored in the specified filename or the inline
+// certificate cert.
+func newCertStore(cert, filename string) (CertGetter, error) {
+	if len(filename) > 0 {
+		return &fileCertStore{filename: filename}, nil
+	} else if len(cert) > 0 {
+		return inlineCertStore{cert: []byte(cert)}, nil
+	} else {
+		return nil, errors.New("invalid certificate inputs")
+	}
+}
+
+// inlineCertStore implements a CertGetter that never changes.
+type inlineCertStore struct {
+	cert []byte
+}
+
+func (s inlineCertStore) GetCert() ([]byte, bool, error) {
+	return s.cert, false, nil
+}
+
+// fileCertStore loads a certificate from a filename each time the
+// certificate is requested.
+type fileCertStore struct {
+	filename string
+	mtx      sync.Mutex // mtx protects accesses to cert
+	cert     []byte
+}
+
+func (s *fileCertStore) GetCert() ([]byte, bool, error) {
+	updated := false
+	newCert, err := readCertificateFile(s.filename)
+	if err != nil {
+		return nil, false, err
+	}
+
+	s.mtx.Lock()
+	if !bytes.Equal(s.cert, newCert) {
+		s.cert = newCert
+		updated = true
+	}
+	s.mtx.Unlock()
+
+	return newCert, updated, nil
+}
+
 // tlsRoundTripper is a RoundTripper that updates automatically its TLS
 // configuration whenever the content of the CA file changes.
 type tlsRoundTripper struct {
-	caFile string
+	certStore CertGetter
 	// newRT returns a new RoundTripper.
 	newRT func(*tls.Config) (http.RoundTripper, error)
 
-	mtx        sync.RWMutex
```

Comment:
This is the change that I feel the most uneasy about because the existing mutex is protecting multiple things: access to the hash, the round tripper and the tlsConfig.